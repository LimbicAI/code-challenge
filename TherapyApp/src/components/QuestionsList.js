import React from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  FlatList,
  View,
  StyleSheet,
  Text,
  TouchableOpacity,
} from 'react-native';

const QuestionsList = ({questions = [], itemOnPress, fetchMore}) => (
  <View style={styles.list}>
    <FlatList
      data={questions}
      keyExtractor={({id}) => `question-${id}`}
      ListEmptyComponent={<Text>Questions list is empty</Text>}
      renderItem={({item: {id, text}}) => (
        <TouchableOpacity
          key={id}
          onPress={() => itemOnPress(id)}
          style={styles.listItem}>
          <Text>{text}</Text>
        </TouchableOpacity>
      )}
    />
    <Button title="Fetch more" onPress={fetchMore} />
  </View>
);

const styles = StyleSheet.create({
  list: {
    borderTopWidth: 1,
    borderColor: 'lightgray',
  },
  listItem: {
    padding: 5,
    borderBottomWidth: 1,
    borderColor: 'lightgray',
  },
});

QuestionsList.propTypes = {
  questions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
    }),
  ),
  itemOnPress: PropTypes.func.isRequired,
  fetchMore: PropTypes.func.isRequired,
};

export default QuestionsList;
