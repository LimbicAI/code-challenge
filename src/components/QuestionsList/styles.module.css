.buttonGroup {
	animation-name: buttonSlide;
	animation-duration: 0.25s;
	min-width: 148px;
}

.modal {
	display: flex;
	flex-flow: column;
	justify-content: center;
	align-items: center;
	background-color: rgba(0, 0, 0, 0.4);
	position: absolute;
	top: 0;
	left: 0;
	width: 100vw;
	height: 100vh;
}

.noQuestions {
	margin: 16px 0;
	font-size: 20px;
}

.question {
	display: flex;
	flex-flow: row;
	justify-content: space-between;
	font-size: 18px;
	font-weight: 500;
	align-items: center;
	width: 90%;
	min-width: 340px;
}

.questionText {
	text-align: left;
	animation-name: questionSlide;
	animation-duration: 0.25s;
}

.questions {
	display: flex;
	flex-flow: column;
	align-items: center;
}

.trashCanIcon {
	height: 16px;
}

@keyframes questionSlide {
	from {
		transform: translateX(-50%);
		opacity: 0;
	}

	to {
		transform: translateX(0);
		opacity: 1;
	}
}

@keyframes buttonSlide {
	from {
		transform: translateX(50%);
		opacity: 0;
	}

	to {
		transform: translateX(0);
		opacity: 1;
	}
}

@media only screen and (min-width: 600px) {
	.question {
		width: 80%;
	}
}

@media only screen and (min-width: 960px) {
	.question {
		width: 60%;
	}
}
