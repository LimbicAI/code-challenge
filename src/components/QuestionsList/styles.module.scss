.buttonGroup {
  animation-name: buttonSlide;
  animation-duration: 0.25s;
  min-width: 148px;
}

.modal {
  display: flex;
  flex-flow: column;
  justify-content: center;
  align-items: center;
  background-color: rgba(0, 0, 0, 0.4);
  position: absolute;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
}

.noQuestions {
  margin: 16px 0;
  font-size: 20px;
}

.question {
  align-items: center;
  display: flex;
  flex-flow: row;
  font-size: 18px;
  font-weight: 500;
  justify-content: space-between;
  margin: 4px 0;
  min-width: 340px;
  width: 90%;
}

.questionText {
  text-align: left;
  animation-name: questionSlide;
  animation-duration: 0.25s;
}

.questions {
  display: flex;
  flex-flow: column;
  align-items: center;
}

.buttonIcon {
  height: 16px;
}

@keyframes questionSlide {
  from {
    transform: translateX(-50%);
    opacity: 0;
  }

  to {
    transform: translateX(0);
    opacity: 1;
  }
}

@keyframes buttonSlide {
  from {
    transform: translateX(50%);
    opacity: 0;
  }

  to {
    transform: translateX(0);
    opacity: 1;
  }
}

@media only screen and (min-width: 600px) {
  .question {
    width: 80%;
  }
}

@media only screen and (min-width: 960px) {
  .question {
    width: 60%;
  }
}