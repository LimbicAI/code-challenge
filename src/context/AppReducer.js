const updateLocalStorage = (type, questions) => {
	localStorage.setItem(type, JSON.stringify(questions));
};

const reducer = (state, action) => {
	let updatedQuestions;
	let updatedClients = [];
	switch (action.type) {
		case 'REMOVE_QUESTION':
			updatedQuestions = state.questions.filter((question, index) => {
				return index !== action.payload.index;
			});

			updatedClients = [];
			state.clients.forEach(client => {
				updatedClients.push({
					name: client.name,
					questions: client.questions.filter(
						question => question.q !== action.payload.description
					)
				});
			});

			updateLocalStorage('questions', updatedQuestions);
			updateLocalStorage('clients', updatedClients);

			return {
				...state,
				questions: updatedQuestions,
				clients: updatedClients
			};
		case 'ADD_QUESTION':
			const { description, choices, type } = action.payload;
			const newQuestion = {
				description,
				choices,
				type
			};
			updatedQuestions = [newQuestion, ...state.questions];
			updatedClients = state.clients.map(client => client);
			updatedClients.forEach(client => {
				client.questions.push({ q: description, a: '-' });
			});

			updateLocalStorage('questions', updatedQuestions);
			updateLocalStorage('clients', updatedClients);

			return {
				...state,
				questions: updatedQuestions
			};
		case 'EDIT_QUESTION':
			const updatedQuestion = {
				id: action.payload.id,
				description: action.payload.description,
				type: action.payload.type,
				choices: action.payload.choices
			};

			updatedQuestions = state.questions.map((question, index) => {
				if (index === updatedQuestion.id) {
					return updatedQuestion;
				}
				return question;
			});

			updatedClients = [];
			state.clients.forEach(client => {
				updatedClients.push({
					name: client.name,
					questions: client.questions.map((question, index) => {
						if (action.payload.oldDescription === question.q) {
							return {
								...question,
								q: action.payload.description
							};
						}
						return question;
					})
				});
			});

			updateLocalStorage('questions', updatedQuestions);
			updateLocalStorage('clients', updatedClients);

			return {
				...state,
				questions: updatedQuestions,
				clients: updatedClients
			};

		case 'UPDATE_ANSWERS':
			const clientName = action.payload.client.name;
			const updatedClient = {
				name: clientName
			};
			updatedClient.questions = Object.values(action.payload.answers).map(
				(answer, index) => {
					return {
						q: action.payload.client.questions[index].q,
						a: answer
					};
				}
			);

			const position = state.clients.findIndex(
				client => client.name === clientName
			);
			updatedClients = [...state.clients];
			updatedClients[position] = updatedClient;
			updateLocalStorage('clients', updatedClients);

			return {
				...state,
				clients: updatedClients
			};

		default:
			return state;
	}
};

export default reducer;
