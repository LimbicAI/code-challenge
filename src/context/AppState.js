import React, { createContext, useReducer } from 'react';
import AppReducer from './AppReducer';
import CLIENTS from '../fixtures/clients';
import QUESTIONS from '../fixtures/questions';

const savedQuestions = localStorage.getItem('questions');
const savedClients = localStorage.getItem('clients');

const initialState = {
	questions: savedQuestions ? JSON.parse(savedQuestions) : QUESTIONS,
	clients: savedClients ? JSON.parse(savedClients) : CLIENTS
};

export const AppContext = createContext(initialState);

export const AppProvider = ({ children }) => {
	const [state, dispatch] = useReducer(AppReducer, initialState);

	const removeQuestion = payload => {
		dispatch({
			type: 'REMOVE_QUESTION',
			payload
		});
	};

	const addQuestion = question => {
		dispatch({
			type: 'ADD_QUESTION',
			payload: question
		});
	};

	const editQuestion = question => {
		dispatch({
			type: 'EDIT_QUESTION',
			payload: question
		});
	};

	const updateAnswers = (client, answers, questions) => {
		dispatch({
			type: 'UPDATE_ANSWERS',
			payload: { client, answers, questions }
		});
	};

	return (
		<AppContext.Provider
			value={{
				clients: state.clients,
				questions: state.questions,
				removeQuestion,
				addQuestion,
				editQuestion,
				updateAnswers
			}}
		>
			{children}
		</AppContext.Provider>
	);
};
