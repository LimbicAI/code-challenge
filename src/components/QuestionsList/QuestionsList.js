import React, { useContext, useState } from 'react';
import { AppContext } from '../../context/AppState';
import QuestionModal from '../QuestionModal';
import Button from '../Button';
import styles from './styles.module.css';

const QuestionsList = () => {
	const { questions, removeQuestion } = useContext(AppContext);
	const [showModal, setShowModal] = useState(false);
	const [modalType, setModalType] = useState('create');
	const [question, setQuestion] = useState({});

	const openCreateQuestion = () => {
		setQuestion({ description: '', index: null });
		setShowModal(true);
		setModalType('create');
	};

	const openEditQuestion = (event, question, index) => {
		event.preventDefault();
		setQuestion({ ...question, id: index });
		setShowModal(true);
		setModalType('edit');
	};

	const toggleModal = value => {
		setShowModal(value);
	};

	return (
		<div className={styles.questions}>
			{questions.map((question, index) => (
				<div
					className={styles.question}
					key={`${question.description}_${index}`}
				>
					<span className={styles.questionText}>
						{question.description}
					</span>
					<div className={styles.buttonGroup}>
						<Button
							onClick={() =>
								removeQuestion({
									index,
									description: question.description
								})
							}
							variation="alert"
						>
							Delete
						</Button>
						<Button
							onClick={event =>
								openEditQuestion(event, question, index)
							}
							variation="small"
						>
							Edit
						</Button>
					</div>
				</div>
			))}
			<Button onClick={openCreateQuestion}>Add a Question</Button>
			{showModal && (
				<QuestionModal
					type={modalType}
					toggleModal={toggleModal}
					value={question}
				/>
			)}
		</div>
	);
};

export default QuestionsList;
