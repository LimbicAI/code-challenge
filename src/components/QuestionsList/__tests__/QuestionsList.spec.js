import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { AppContext, AppProvider } from '../../../context/AppState';
import QuestionsList from '../QuestionsList';
import QUESTIONS from '../../../fixtures/questions';

let result;

const setup = data => {
	result = render(
		<AppProvider>
			<QuestionsList />
		</AppProvider>
	);
};

describe('QuestionList tests', () => {
	it('matches the snapshot', () => {
		setup();
		expect(result.asFragment()).toMatchSnapshot();
	});

	it('has no questions copy when no questions', () => {
		setup([]);
		render(
			<AppContext.Provider value={{ questions: [] }}>
				<QuestionsList />
			</AppContext.Provider>
		);
		screen.debug();
		expect(
			screen.queryByText(/no questions available/i)
		).toBeInTheDocument();
	});

	it('opens the modal on Add a Question click', () => {
		setup();
		expect(screen.queryByText(/create/i)).not.toBeInTheDocument();
		expect(screen.queryByText(/cancel/i)).not.toBeInTheDocument();
		fireEvent.click(screen.getByText(/add a question/i));
		expect(screen.queryByText(/create/i)).toBeInTheDocument();
		expect(screen.queryByText(/cancel/i)).toBeInTheDocument();
	});

	it('creates a Question on Create click and form submission', () => {
		setup();
		const newQuestion = 'Is this a test question?';
		expect(screen.queryByText(newQuestion)).not.toBeInTheDocument();
		fireEvent.click(screen.getByText(/add a question/i));
		fireEvent.change(screen.getByLabelText(/question/i), {
			target: { value: newQuestion }
		});
		fireEvent.click(screen.getByText(/create/i));
		expect(screen.queryByText(newQuestion)).toBeInTheDocument();
	});

	it('opens the modal on Edit click', () => {
		setup();
		expect(screen.queryByText(/update/i)).not.toBeInTheDocument();
		expect(screen.queryByText(/cancel/i)).not.toBeInTheDocument();
		fireEvent.click(screen.queryAllByTestId('pencil-icon')[0]);
		expect(screen.queryByText(/update/i)).toBeInTheDocument();
		expect(screen.queryByText(/cancel/i)).toBeInTheDocument();
	});

	it('updates a Question on Create click and form submission', () => {
		setup();
		const updatedQuestion = 'Is this a test question?';
		expect(
			screen.queryByText(QUESTIONS[0].description)
		).toBeInTheDocument();
		expect(screen.queryByText(updatedQuestion)).not.toBeInTheDocument();
		fireEvent.click(screen.queryAllByTestId('pencil-icon')[0]);
		fireEvent.change(screen.getByLabelText(/question/i), {
			target: { value: updatedQuestion }
		});
		fireEvent.click(screen.getByText(/update/i));
		expect(
			screen.queryByText(QUESTIONS[0].description)
		).not.toBeInTheDocument();
		expect(screen.queryByText(updatedQuestion)).toBeInTheDocument();
	});

	it('deletes a question on Delete click', () => {
		setup();
		expect(
			screen.queryByText(QUESTIONS[0].description)
		).toBeInTheDocument();
		fireEvent.click(screen.queryAllByTestId('trash-can-icon')[0]);
		expect(
			screen.queryByText(QUESTIONS[0].description)
		).not.toBeInTheDocument();
	});
});
