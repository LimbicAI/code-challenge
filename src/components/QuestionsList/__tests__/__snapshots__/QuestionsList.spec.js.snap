// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="questions"
  >
    <div
      class="question"
    >
      <span
        class="questionText"
      >
        How are you feeling today?
      </span>
      <div
        class="buttonGroup"
      >
        <button
          class="button alert"
          type="button"
        >
          Delete
        </button>
        <button
          class="button small"
          type="button"
        >
          Edit
        </button>
      </div>
    </div>
    <div
      class="question"
    >
      <span
        class="questionText"
      >
        How stressed are you feeling today?
      </span>
      <div
        class="buttonGroup"
      >
        <button
          class="button alert"
          type="button"
        >
          Delete
        </button>
        <button
          class="button small"
          type="button"
        >
          Edit
        </button>
      </div>
    </div>
    <div
      class="question"
    >
      <span
        class="questionText"
      >
        What feelings best describe your mood today?
      </span>
      <div
        class="buttonGroup"
      >
        <button
          class="button alert"
          type="button"
        >
          Delete
        </button>
        <button
          class="button small"
          type="button"
        >
          Edit
        </button>
      </div>
    </div>
    <button
      class="button default"
      type="button"
    >
      Add a Question
    </button>
  </div>
</DocumentFragment>
`;
